from PyQt6.QtCore import pyqtSignal, Qt
from PyQt6.QtGui import QPainter
from PyQt6.QtWidgets import QWidget, QHBoxLayout, QLabel, QListWidgetItem, QMessageBox

from .base_page import BasePage
from simple_api_client import SimpleTheatreClient  # –î–æ–±–∞–≤–ª–µ–Ω –∏–º–ø–æ—Ä—Ç


class HomePage(BasePage):
        navigate_to = pyqtSignal(str)

        def __init__(self, parent=None):
            super().__init__("ui/home_page.ui", parent)
            self.client = SimpleTheatreClient()  # –î–æ–±–∞–≤–ª–µ–Ω –∫–ª–∏–µ–Ω—Ç API
            self.setup_displays()
            self.setup_export()

        def showEvent(self, event):
            self.setup_displays()
            super().showEvent(event)

        def setup_displays(self):
            # –ü–æ–ª—É—á–∞–µ–º –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏—é –æ –±–ª–∏–∂–∞–π—à–µ–º –∑–∞–Ω—è—Ç–∏–∏ —á–µ—Ä–µ–∑ API
            lessons = ""
            lessons = self.client.get_lessons()
            if lessons:
                # –ë–µ—Ä–µ–º –ø–µ—Ä–≤–æ–µ –∑–∞–Ω—è—Ç–∏–µ (–ø—Ä–µ–¥–ø–æ–ª–∞–≥–∞–µ–º, —á—Ç–æ API –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç –æ—Ç—Å–æ—Ä—Ç–∏—Ä–æ–≤–∞–Ω–Ω—ã–π —Å–ø–∏—Å–æ–∫)
                lesson = lessons[0]
                title = lesson['title']
                time = lesson['time']
                date = lesson['date']

                # –ü–∞—Ä—Å–∏–º –¥–∞—Ç—É
                day, month, year = date.split("-")
                months = [
                    "—è–Ω–≤–∞—Ä—è", "—Ñ–µ–≤—Ä–∞–ª—è", "–º–∞—Ä—Ç–∞", "–∞–ø—Ä–µ–ª—è", "–º–∞—è", "–∏—é–Ω—è",
                    "–∏—é–ª—è", "–∞–≤–≥—É—Å—Ç–∞", "—Å–µ–Ω—Ç—è–±—Ä—è", "–æ–∫—Ç—è–±—Ä—è", "–Ω–æ—è–±—Ä—è", "–¥–µ–∫–∞–±—Ä—è"
                ]

                date_text = """
                <div align="center" style="line-height: 0.7;">
                    <span style="
                        font-size: 48pt;
                        font-weight: bold;
                        color: #fdefc8;
                        font-family: 'Martian Mono Condensed';
                    ">{day}</span><br>

                    <span style="
                        font-size: 12pt;
                        color: #fdefc8;
                        font-family: 'Martian Mono Condensed';
                    ">{month}</span><br>

                    <span style="
                        font-size: 16pt;
                        color: #fdefc8;
                        font-family: 'Martian Mono Condensed';
                    ">{time}</span><br>
                </div>
                """

                lesson_text = """
                <div style="
                    font-family: 'Oranienbaum';
                    font-size: 42pt;
                    color: #fdefc8;
                    text-align: left;
                    line-height: 1;
                ">
                    {title}<br>
                </div>
                """

                self.dateDisplay.setHtml(date_text.format(
                    day=day, month=months[int(month) - 1], time=time))
                self.lessonDisplay.setHtml(lesson_text.format(title=title))

            self.listWidget.clear()

            # –ü–æ–ª—É—á–∞–µ–º –æ—Ä–≥–∞–Ω–∏–∑–∞—Ç–æ—Ä–æ–≤ —á–µ—Ä–µ–∑ API
            organizers = self.client.get_organizers()
            for organizer in organizers:
                name = organizer['username']
                avatar_data = organizer.get('avatar')
                custom_item = CustomListItem(name, avatar_data)
                item = QListWidgetItem()
                item.setSizeHint(custom_item.sizeHint())
                self.listWidget.addItem(item)
                self.listWidget.setItemWidget(item, custom_item)

            try:
                participants = self.client.get_participants()
                self.participantsList.clear()
                
                for participant in participants:
                    item = f"{participant['username']}"
                    self.participantsList.addItem(item)
                        
            except Exception as e:
                print(f"–û—à–∏–±–∫–∞ –∑–∞–≥—Ä—É–∑–∫–∏ —É—á–∞—Å—Ç–Ω–∏–∫–æ–≤: {e}")

        def setup_export(self):
            """–ù–∞—Å—Ç—Ä–æ–π–∫–∞ —ç–∫—Å–ø–æ—Ä—Ç–∞"""
            self.exportButton.clicked.connect(self.export_to_csv)

        def export_to_csv(self):
            """–≠–∫—Å–ø–æ—Ä—Ç –¥–∞–Ω–Ω—ã—Ö –≤ CSV"""
            try:
                from datetime import datetime
                import csv
                
                # –ü–æ–ª—É—á–∞–µ–º –¥–∞–Ω–Ω—ã–µ –¥–ª—è —ç–∫—Å–ø–æ—Ä—Ç–∞
                data = self.get_export_data()
                if not data:
                    QMessageBox.warning(self, "–û—à–∏–±–∫–∞", "–ù–µ—Ç –¥–∞–Ω–Ω—ã—Ö –¥–ª—è —ç–∫—Å–ø–æ—Ä—Ç–∞")
                    return
                    
                # –°–æ–∑–¥–∞–µ–º –∏–º—è —Ñ–∞–π–ª–∞
                timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
                filename = f"export_{self.__class__.__name__}_{timestamp}.csv"
                
                # –°–æ—Ö—Ä–∞–Ω—è–µ–º –≤ CSV
                with open(filename, 'w', newline='', encoding='utf-8') as file:
                    if data:
                        writer = csv.DictWriter(file, fieldnames=data[0].keys())
                        writer.writeheader()
                        writer.writerows(data)
                        
                QMessageBox.information(self, "–£—Å–ø–µ—Ö", f"–î–∞–Ω–Ω—ã–µ —ç–∫—Å–ø–æ—Ä—Ç–∏—Ä–æ–≤–∞–Ω—ã –≤ {filename}")
                
            except Exception as e:
                QMessageBox.warning(self, "–û—à–∏–±–∫–∞", f"–û—à–∏–±–∫–∞ —ç–∫—Å–ø–æ—Ä—Ç–∞: {str(e)}")

        def get_export_data(self):
            return [{"participant": item.text().replace("üë§ ", "")} 
                    for item in range(self.participantsList.count())]
        
class CustomListItem(QWidget):
    def __init__(self, name, avatar_data):
        super().__init__()

        layout = QHBoxLayout()

        # –°–æ–∑–¥–∞–µ–º –º–µ—Ç–∫—É –¥–ª—è –æ—Ç–æ–±—Ä–∞–∂–µ–Ω–∏—è –∏–º–µ–Ω–∏
        label = QLabel(name)
        label.setStyleSheet(
            "font-family: 'Martian Mono Condensed'; color: #fdefc8; font-size: 14px;")

        # –°–æ–∑–¥–∞–µ–º –º–µ—Ç–∫—É –¥–ª—è –æ—Ç–æ–±—Ä–∞–∂–µ–Ω–∏—è –∞–≤–∞—Ç–∞—Ä–∞
        mate = QLabel()

        try:
            if avatar_data:
                from PyQt6.QtGui import QPixmap
                from PyQt6.QtCore import QByteArray

                pixmap = QPixmap()

                # –ï—Å–ª–∏ avatar_data - —ç—Ç–æ base64 —Å—Ç—Ä–æ–∫–∞
                if isinstance(avatar_data, str):
                    # –î–µ–∫–æ–¥–∏—Ä—É–µ–º base64 —Å—Ç—Ä–æ–∫—É –≤ –±–∞–π—Ç—ã
                    avatar_bytes = QByteArray.fromBase64(avatar_data.encode())
                    success = pixmap.loadFromData(avatar_bytes)
                else:
                    # –ï—Å–ª–∏ —ç—Ç–æ —É–∂–µ –±–∞–π—Ç—ã
                    success = pixmap.loadFromData(avatar_data)

                if success and not pixmap.isNull():
                    # –ú–∞—Å—à—Ç–∞–±–∏—Ä—É–µ–º –∏ –¥–µ–ª–∞–µ–º –∫—Ä—É–≥–ª—ã–º
                    pixmap = pixmap.scaled(
                        60,
                        60,
                        Qt.AspectRatioMode.IgnoreAspectRatio,
                        Qt.TransformationMode.SmoothTransformation)
                    rounded = QPixmap(pixmap.size())
                    rounded.fill(Qt.GlobalColor.transparent)

                    # –°–æ–∑–¥–∞–µ–º –∫—Ä—É–≥–ª—É—é –º–∞—Å–∫—É
                    rounded = QPixmap(60, 60)
                    rounded.fill(Qt.GlobalColor.transparent)

                    painter = QPainter(rounded)
                    painter.setRenderHint(QPainter.RenderHint.Antialiasing)

                    # –°–æ–∑–¥–∞–µ–º –∫—Ä—É–≥–ª—ã–π –∫–ª–∏–ø
                    painter.setBrush(Qt.GlobalColor.black)
                    painter.drawEllipse(0, 0, 60, 60)

                    # –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º –∫–æ–º–ø–æ–∑–∏—Ü–∏–æ–Ω–Ω—ã–π —Ä–µ–∂–∏–º –¥–ª—è –æ–±—Ä–µ–∑–∫–∏
                    painter.setCompositionMode(
                        QPainter.CompositionMode.CompositionMode_SourceIn)

                    # –†–∏—Å—É–µ–º –∏–∑–æ–±—Ä–∞–∂–µ–Ω–∏–µ –≤–Ω—É—Ç—Ä–∏ –∫—Ä—É–≥–∞
                    painter.drawPixmap(0, 0, pixmap)
                    painter.end()

                    mate.setPixmap(rounded)
                    mate.setText("")
                else:
                    mate.setText("üë§")  # –ò–∫–æ–Ω–∫–∞ –≤–º–µ—Å—Ç–æ —Ç–µ–∫—Å—Ç–∞
            else:
                mate.setText("üë§")  # –ò–∫–æ–Ω–∫–∞ –µ—Å–ª–∏ –Ω–µ—Ç –∞–≤–∞—Ç–∞—Ä–∞

        except Exception as e:
            print(f"Error loading avatar for {name}: {e}")
            mate.setText("üë§")  # –ò–∫–æ–Ω–∫–∞ –ø—Ä–∏ –æ—à–∏–±–∫–µ

        layout.addWidget(mate)
        layout.addWidget(label)
        layout.addStretch()

        self.setLayout(layout)
